File: python/FiveCardStud.py
import random
from collections import defaultdict
from enum import Enum
from Deck import Deck
from Card import Card
from functools import cmp_to_key

# enum for suits
class Suit(Enum):
    DIAMONDS = 'D'
    CLUBS = 'C'
    HEARTS = 'H'
    SPADES = 'S'

# to rank suits Diamonds < Clubs < Hearts < Spades
SUIT_RANKING = {
    'D': 1,  
    'C': 2,  
    'H': 3,  
    'S': 4   
}

# get suit rank based on the suit string
def suit_rank(suit):
    return SUIT_RANKING[suit]

# enum for hand ranks
class HandRank(Enum):
    HIGH_CARD = 1
    PAIR = 2
    TWO_PAIR = 3
    THREE_OF_A_KIND = 4
    STRAIGHT = 5
    FLUSH = 6
    FULL_HOUSE = 7
    FOUR_OF_A_KIND = 8
    STRAIGHT_FLUSH = 9
    ROYAL_FLUSH = 10

    # comparison operators for hand ranks based on their value
    def __lt__(self, other):
        return self.value < other.value

    def __gt__(self, other):
        return self.value > other.value

# convert hand rank enum to string
def hand_rank_to_string(rank):
    if rank == HandRank.HIGH_CARD:
        return "High Card"
    elif rank == HandRank.PAIR:
        return "Pair"
    elif rank == HandRank.TWO_PAIR:
        return "Two Pair"
    elif rank == HandRank.THREE_OF_A_KIND:
        return "Three Of A Kind"
    elif rank == HandRank.STRAIGHT:
        return "Straight"
    elif rank == HandRank.FLUSH:
        return "Flush"
    elif rank == HandRank.FULL_HOUSE:
        return "Full House"
    elif rank == HandRank.FOUR_OF_A_KIND:
        return "Four Of A Kind"
    elif rank == HandRank.STRAIGHT_FLUSH:
        return "Straight Flush"
    elif rank == HandRank.ROYAL_FLUSH:
        return "Royal Straight Flush"
    else:
        return "Unknown"

# method to handle command line args 
def read_deck_from_file(filename, hands):
    seen_cards = set() # to track duplicate cards

    try:
        with open(filename, 'r') as file:
            lines = file.readlines()  
            
            for line in lines:
                print(line.strip())  # display file line to user

            for i in range(6):
                card_strings = lines[i].strip().split(",")
                for card_str in card_strings:
                    card_str = card_str.strip() 
                    
                    # duplicate card check
                    if card_str in seen_cards:
                        print("\n*** ERROR - DUPLICATED CARD FOUND IN DECK ***")
                        print(f"\n\n*** DUPLICATE: {card_str} ***")
                        return  
                    
                    seen_cards.add(card_str)  
                    suit = card_str[-1]  
                    rank_str = card_str[:-1]

                    hands[i].append(Card(rank_str, suit))  

            print()
            print_six_hands(hands)
            
    except FileNotFoundError:
        print(f"Error: The file {filename} was not found.")
    except Exception as e:
        print(f"An error occurred: {e}")

    return hands

# method to print winning hand 
def print_winning_hand(hands, hand_ranks):
    print("--- WINNING HAND ORDER ---")
    for i in range(len(hands)):
        hand = hands[i]
        hand_string = " ".join([f"{Card.rank_to_string(card.rank)}{card.suit}" for card in hand])
        print(f"{hand_string} - {hand_rank_to_string(hand_ranks[i])}")
    print()  

# method to print six hands 
def print_six_hands(hands):
    print("*** Here are the six hands...")
    for i in range(6):  
        hand = hands[i]  
        card_strings = [f"{Card.rank_to_string(card.rank)}{card.suit}" for card in hand]
        print(" ".join(card_strings))  
    print()  

# method to compare two hands and determine the winner
def compare_hands(hand1, hand2):
    rank1 = classify_hand(hand1)
    rank2 = classify_hand(hand2)
    
    if rank1 != rank2:
        return -1 if rank1 < rank2 else 1

    # tie breaking logic

    # case 1: Flush 
    if rank1 in {HandRank.FLUSH, HandRank.STRAIGHT_FLUSH, HandRank.ROYAL_FLUSH}:
        highest_card1 = get_highest_card(hand1, False)
        highest_card2 = get_highest_card(hand2, False)

        if highest_card1.rank != highest_card2.rank:
            if highest_card1.rank < highest_card2.rank:
                return -1
            else:
                return  1
        if suit_rank(highest_card1.suit) < suit_rank(highest_card2.suit):
            return -1
        else:
            return 1

    # case 2: Straight
    if rank1 == HandRank.STRAIGHT:
        check_ace1 = is_ace_low_straight(hand1)
        check_ace2 = is_ace_low_straight(hand2)

        highest_card1 = get_highest_card(hand1, check_ace1)
        highest_card2 = get_highest_card(hand2, check_ace2)

        if highest_card1.rank != highest_card2.rank:
            if highest_card1.rank < highest_card2.rank:
                return -1
            else: 
                return 1
        if suit_rank(highest_card1.suit) < suit_rank(highest_card2.suit): 
            return -1
        else: 
            return 1

    # case 3: Two Pair 
    if rank1 == HandRank.TWO_PAIR:
        high_pair1 = get_highest_pair_rank(hand1)
        high_pair2 = get_highest_pair_rank(hand2)

        if high_pair1 != high_pair2:
            if high_pair1 < high_pair2: 
                return -1
            else:
                return 1

        low_pair1 = get_lowest_pair_rank(hand1)
        low_pair2 = get_lowest_pair_rank(hand2)

        if low_pair1 != low_pair2:
            if low_pair1 < low_pair2:
                return -1
            else:
                return 1

        kicker1 = get_kicker(hand1)
        kicker2 = get_kicker(hand2)
        if suit_rank(hand1[kicker1].suit) < suit_rank(hand2[kicker2].suit):
            return -1 
        else:
            return 1

    # case 4: Pair 
    if rank1 == HandRank.PAIR:
        pair_rank1 = get_pair_rank(hand1)
        pair_rank2 = get_pair_rank(hand2)

        if pair_rank1 != pair_rank2:
            if pair_rank1 < pair_rank2:
                return -1 
            else:
                return 1

        high_val1 = get_highest_non_pair_card(hand1)
        high_val2 = get_highest_non_pair_card(hand2)

        if high_val1.rank != high_val2.rank:
            if high_val1.rank < high_val2.rank: 
                return -1
            else:
                return 1
        if suit_rank(high_val1.suit) < suit_rank(high_val2.suit):
                return -1
        else:
                return 1

    # case 5: high card 
    if rank1 == HandRank.HIGH_CARD:
        highest_card1 = get_highest_card(hand1, False)
        highest_card2 = get_highest_card(hand2, False)

        if highest_card1.rank != highest_card2.rank:
            if highest_card1.rank < highest_card2.rank:
                return -1
            else:
                return 1
        if suit_rank(highest_card1.suit) < suit_rank(highest_card2.suit):
            return -1
        else:
            return 1
    
    # handle tiebreakders for general poker rules 
    # three of a kind, four of a kind, full house     

    if rank1 == HandRank.FOUR_OF_A_KIND:
        quad_rank1 = get_quad_rank(hand1)
        quad_rank2 = get_quad_rank(hand2)
        if quad_rank1 < quad_rank2:
            return -1
        else:
            return 1

    if rank1 == HandRank.THREE_OF_A_KIND:
        triplet_rank1 = get_triplet_rank(hand1)
        triplet_rank2 = get_triplet_rank(hand2)
        if triplet_rank1 < triplet_rank2: 
            return -1
        else:
            return 1

    if rank1 == HandRank.FULL_HOUSE:
        triplet_rank1 = get_triplet_rank(hand1)
        triplet_rank2 = get_triplet_rank(hand2)

        if triplet_rank1 != triplet_rank2:
            if triplet_rank1 < triplet_rank2:
                return -1
            else:
                return 1

        pair_rank1 = get_pair_rank(hand1)
        pair_rank2 = get_pair_rank(hand2)
        if pair_rank1 < pair_rank2:
            return -1
        else:
            return 1

    return 0

# checks if the hand is a straight
def is_straight(hand):
    ranks = set()

    for card in hand:
        ranks.add(card.rank)

    sorted_ranks = sorted(ranks)  # Ascending order

    # Special case for Ace-low straight
    if sorted_ranks == [2, 3, 4, 5, 14]:
        return True

    for i in range(len(sorted_ranks) - 1):
        if sorted_ranks[i + 1] - sorted_ranks[i] != 1:
            return False

    return True

# classifies the rank of the hand
def classify_hand(hand):
    hand_copy = hand[:] # shallow copy of hand
    hand_copy.sort(key=lambda card: card.rank, reverse=True)
    occurrences = [0] * 15
    suits = defaultdict(int)
    
    for card in hand_copy:
        occurrences[card.rank] += 1
        suits[card.suit] += 1
    
    flush = any(count == 5 for count in suits.values())
    straight = is_straight(hand_copy)
    
    if flush and straight:
        royal_ranks = {10, 11, 12, 13, 14}  # Royal flush ranks
        unique_ranks = {card.rank for card in hand_copy}  # Unique ranks in the hand

        if unique_ranks == royal_ranks:
            return HandRank.ROYAL_FLUSH
        return HandRank.STRAIGHT_FLUSH        

    three_count = 0
    pair_count = 0

    for count in occurrences[2:]:
        if count == 4:
            return HandRank.FOUR_OF_A_KIND
        if count == 3:
            three_count += 1
        if count == 2:
            pair_count += 1

    if three_count == 1 and pair_count == 1:
        return HandRank.FULL_HOUSE
    if flush:
        return HandRank.FLUSH
    if straight:
        return HandRank.STRAIGHT
    if three_count == 1:
        return HandRank.THREE_OF_A_KIND
    if pair_count == 2:
        return HandRank.TWO_PAIR
    if pair_count == 1:
        return HandRank.PAIR

    return HandRank.HIGH_CARD

# sorts hands based on their rank using the compare_hands function
def sort_hands(hands):
    hand_ranks = []
    hands.sort(key=cmp_to_key(compare_hands), reverse=True)
    for hand in hands:
        hand_rank = classify_hand(hand)
        hand_ranks.append(hand_rank)
    return hand_ranks

# checks if hand is an Ace Low Straight (A 2 3 4 5)
def is_ace_low_straight(hand):
    ranks = {card.rank for card in hand}
    return ranks == {14, 2, 3, 4, 5}   

# returns the highest card in the hand
# considers case of Ace Low Straight
def get_highest_card(hand, is_ace_straight):
    temp_hand = hand.copy()
    if is_ace_straight:
        for i in range(len(temp_hand)):
            if temp_hand[i].rank == 14:
                temp_hand[i].rank = 1  # Treat Ace as low

    highest_card = temp_hand[0]
    for i in range(1, len(temp_hand)):
        if temp_hand[i].rank > highest_card.rank:
            highest_card = temp_hand[i]
    return highest_card

# counts the number of times a specific rank occurs in the hand
def count_occurrences(hand, rank):
    count = 0  
    for card in hand:
        if card.rank == rank:
            count += 1  
    return count

# returns the highest non-pair card (the kicker)
def get_highest_non_pair_card(hand):
    temp_hand = []  

    for card in hand:
        if count_occurrences(hand, card.rank) != 2:  
            temp_hand.append(card)

    if not temp_hand:  
        return None  

    highest_card = temp_hand[0] 
    for card in temp_hand[1:]:  
        if card.rank > highest_card.rank:  
            highest_card = card  

    return highest_card  

# identifies and returns the index of the kicker card in the hand
def get_kicker(hand):
    max_val = 0
    kicker_index = -1

    for index, card in enumerate(hand): 
        if count_occurrences(hand, card.rank) == 1 and card.rank > max_val:
            max_val = card.rank
            kicker_index = index

    return kicker_index

# get the rank (value) of the pair in a hand
def get_pair_rank(hand):
    for card in hand:
        if count_occurrences(hand, card.rank) == 2:  
            return card.rank
    return -1  

# get the highest rank (value) of pairs in a hand
def get_highest_pair_rank(hand):
    highest_pair = -1
    for card in hand:
        if count_occurrences(hand, card.rank) == 2 and card.rank > highest_pair:
            highest_pair = card.rank  
    return highest_pair  

# get the lowest rank (value) of pairs in a hand
def get_lowest_pair_rank(hand):
    lowest_pair = 15  
    for card in hand:
        if count_occurrences(hand, card.rank) == 2 and card.rank < lowest_pair:
            lowest_pair = card.rank  

    if lowest_pair == 15:
        return -1  
    return lowest_pair  

# get the rank of the three-of-a-kind in a hand
def get_triplet_rank(hand):
    for card in hand:
        if count_occurrences(hand, card.rank) == 3: 
            return card.rank  
    return -1  

# get the rank of the four-of-a-kind in a hand
def get_quad_rank(hand):
    for card in hand:
        if count_occurrences(hand, card.rank) == 4: 
            return card.rank  
    return -1  

# main method to handle overall program flow 
def main(args):
    deck = Deck()
    hands = [[] for _ in range(6)]  # create 6 empty hands
     
    print("*** P O K E R    H A N D    A N A L Y Z E R ***\n")
    print()

    if len(args) > 0:  # part 2 with command line args
        print("*** USING TEST DECK ***")
        print(f"\n*** File: {args[0]}")
        if read_deck_from_file(args[0], hands) is None:
            return  # stop execution if a duplicate card was found
    else:  # part 1 without command line args
        print("*** USING RANDOMIZED DECK OF CARDS ***")
        print("\n*** Shuffled 52 card deck:")
        deck.shuffle()
        deck.print_deck(False)
        deck.deal_hands(hands)
        print_six_hands(hands)
        print("*** Here is what remains in the deck...")
        deck.print_deck(True)
        print()

    hand_ranks = sort_hands(hands)
    print_winning_hand(hands, hand_ranks)

if __name__ == "__main__":
    import sys
    main(sys.argv[1:]) # pass comand line args to main

File: perl/FiveCardStud.pl
use strict;
use warnings;
use lib '.'; 
use List::Util qw(shuffle); # for sortHands
use Deck;

# Suit constants and ranking
use constant {
	DIAMONDS => 'D',
    	CLUBS    => 'C',
    	HEARTS   => 'H',
    	SPADES   => 'S',
};

my %SUIT_RANKING = (
    	DIAMONDS => 1,
    	CLUBS    => 2,
    	HEARTS   => 3,
    	SPADES   => 4,
);

# Hand rank constants
use constant {
    	HIGH_CARD       => 1,
    	PAIR            => 2,
    	TWO_PAIR        => 3,
    	THREE_OF_A_KIND => 4,
    	STRAIGHT        => 5,
    	FLUSH           => 6,
    	FULL_HOUSE      => 7,
    	FOUR_OF_A_KIND  => 8,
    	STRAIGHT_FLUSH  => 9,
    	ROYAL_FLUSH     => 10,
};

our @HAND_RANKS = (
    	HIGH_CARD,
    	PAIR,
    	TWO_PAIR,
    	THREE_OF_A_KIND,
    	STRAIGHT,
    	FLUSH,
    	FULL_HOUSE,
    	FOUR_OF_A_KIND,
    	STRAIGHT_FLUSH,
    	ROYAL_FLUSH,
);

# Convert hand rank to string
sub hand_rank_to_string {
    	my ($rank) = @_;
    	return "High Card"       	if $rank == HIGH_CARD;
    	return "Pair"            	if $rank == PAIR;
    	return "Two Pair"        	if $rank == TWO_PAIR;
    	return "Three Of A Kind" 	if $rank == THREE_OF_A_KIND;
    	return "Straight"        	if $rank == STRAIGHT;
    	return "Flush"           	if $rank == FLUSH;
    	return "Full House"      	if $rank == FULL_HOUSE;
    	return "Four Of A Kind"  	if $rank == FOUR_OF_A_KIND;
    	return "Straight Flush"  	if $rank == STRAIGHT_FLUSH;
    	return "Royal Straight Flush"   if $rank == ROYAL_FLUSH;
    	return "Unknown";
}

# main method to handle overall program flow 
sub main {
	my @args = @ARGV;  # Get command-line arguments
	my $deck = Deck->new();
	my @hands = ([], [], [], [], [], []);

	print("*** P O K E R    H A N D    A N A L Y Z E R ***\n\n\n");
	
	if (@args) {  # Part 2: Using a test deck with command-line arguments
		my $filename = $args[0];
        	print "*** USING TEST DECK ***\n";
        	print "\n*** File: $filename\n";
		if (!read_deck_from_file($filename, \@hands)) {
			return;
		}
	} else { # Part 1: Randomized deck without command-line arguments
		print("*** USING RANDOMIZED DECK OF CARDS ***\n\n");
		$deck->shuffle_deck();
		print("*** Shuffled 52 card deck:\n");
		$deck->print_deck(0); 
		$deck->deal_hands(\@hands);
		print_six_hands(\@hands);
		print("*** Here is what remains in the deck...\n");
		$deck->print_deck(1);
		print("\n");
	}

	my $hand_ranks = sort_hands(\@hands);
	print_winning_hand(\@hands, $hand_ranks);
}

main();	

# to print 6 hands of 5 cards 
sub print_six_hands {
    	my ($hands) = @_;
    	print "*** Here are the six hands...\n";
    
    	for my $i (0..5) {
        	my $hand = $hands->[$i];
        	my @card_strings = map { $_->to_string() } @$hand;
        	print join(" ", @card_strings) . "\n";
    	}
    
    	print "\n";
}

# to print the winning hand order based on hand ranks
sub print_winning_hand {
    	my ($hands, $hand_ranks) = @_;
	print "--- WINNING HAND ORDER ---\n";
    
	for my $i (0 .. $#$hands) {
        	my $hand = $hands->[$i];
        	my @card_strings = map { $_->to_string() } @$hand;
        	my $hand_string = join(" ", @card_strings);
        	print "$hand_string - " . hand_rank_to_string($hand_ranks->[$i]) . "\n";
    	}
    	print "\n"; 
}

# method to handle command line args
sub read_deck_from_file {
    	my ($filename, $hands) = @_;
    	my %seen_cards;

	open(my $file, '<', $filename) or do {
        	print "Error: The file $filename was not found.\n";
        	return;
    	};

    	my @lines = <$file>;
    	close $file;

    	foreach my $line (@lines) {
        	print $line; # display line to user 
	}

	for my $i (0 .. 5) {
        	my @card_strings = split /,/, $lines[$i];
        	foreach my $card_str (@card_strings) {
            		$card_str =~ s/^\s+|\s+$//g;	# trim whitespace
			
			# Check for duplicate card
			if ($seen_cards{$card_str}) {
                		print "\n*** ERROR - DUPLICATED CARD FOUND IN DECK ***\n";
                		print "\n\n*** DUPLICATE: $card_str ***\n";
                		return;
            		}
            
            		$seen_cards{$card_str} = 1;
            		my $suit = substr($card_str, -1);
            		my $rank_str = substr($card_str, 0, -1);

            		push @{$hands->[$i]}, Card->new($rank_str, $suit);
        	}
    	}

    	print "\n";
    	print_six_hands($hands);
    
    	return $hands;
}

# sorts hands based on their rank using the compare_hands function
sub sort_hands {
	my ($hands) = @_;
    	my @hand_ranks;

	@$hands = sort { compare_hands($b, $a) } @$hands;
	foreach my $hand (@$hands) {
		my $hand_rank = classify_hand($hand);
		push @hand_ranks, $hand_rank;
	}

	return \@hand_ranks;
}

# determines ie the hand is a straight (considers Ace-Low case)
sub is_straight {
    	my ($hand) = @_;
    	my %ranks;

    	foreach my $card (@$hand) {
        	$ranks{$card->{rank}} = 1;
    	}

    	my @sorted_ranks = sort { $a <=> $b } keys %ranks;

    	return 1 if @sorted_ranks == 5 && join(",", @sorted_ranks) eq "2,3,4,5,14";

    	for my $i (0 .. $#sorted_ranks - 1) {
        	return 0 if $sorted_ranks[$i + 1] - $sorted_ranks[$i] != 1;
    	}

    	return 1;
}

# classifies the rank of the hand
sub classify_hand {
    	my ($hand) = @_;
    	my @hand_copy = sort { $b->{rank} <=> $a->{rank} } @$hand;
    	my @occurrences = (0) x 15;
    	my %suits;

    	foreach my $card (@hand_copy) {
        	$occurrences[$card->{rank}]++;
        	$suits{$card->{suit}}++;
    	}

    	my $flush = grep { $_ == 5 } values %suits;
    	my $straight = is_straight(\@hand_copy);

    	if ($flush && $straight) {
        	my %unique_ranks = map { $_->{rank} => 1 } @hand_copy;
        	return ROYAL_FLUSH if join(",", sort { $a <=> $b } keys %unique_ranks) eq "10,11,12,13,14";
        	return STRAIGHT_FLUSH;
    	}

    	my ($three_count, $pair_count) = (0, 0);

    	foreach my $count (@occurrences[2 .. $#occurrences]) {
        	$three_count++ if $count == 3;
        	$pair_count++ if $count == 2;
        	return FOUR_OF_A_KIND if $count == 4;
    	}

    	return FULL_HOUSE      if $three_count == 1 && $pair_count == 1;
    	return FLUSH           if $flush;
    	return STRAIGHT        if $straight;
    	return THREE_OF_A_KIND if $three_count == 1;
    	return TWO_PAIR        if $pair_count == 2;
    	return PAIR            if $pair_count == 1;

    	return HIGH_CARD;
}

# Checks if a hand is an Ace-Low Straight (A, 2, 3, 4, 5)
sub is_ace_low_straight {
    	my ($hand) = @_;
    	my %ranks = map { $_->{rank} => 1 } @$hand;
    	return exists $ranks{14} && exists $ranks{2} && exists $ranks{3} && exists $ranks{4} && exists $ranks{5};
}

# Returns the highest card in the hand, considering the Ace-Low Straight case
sub get_highest_card {
	my ($hand, $is_ace_straight) = @_;
	my @temp_hand = map { { %$_ } } @$hand;  # Copy hand to avoid modifying the original

	if ($is_ace_straight) {
        	foreach my $card (@temp_hand) {
            		$card->{rank} = 1 if $card->{rank} == 14;
        	}
    	}

    	my $highest_card = $temp_hand[0];
    	for my $i (1 .. $#temp_hand) {
        	if ($temp_hand[$i]{rank} > $highest_card->{rank}) {
            		$highest_card = $temp_hand[$i];
        	}
    	}
    	return $highest_card;
}

# Counts the number of occurrences of a specific rank in the hand
sub count_occurrences {
    	my ($hand, $rank) = @_;
    	my $count = 0;
    	foreach my $card (@$hand) {
        	$count++ if $card->{rank} == $rank;
    	}
    	return $count;
}

# Returns the highest non-pair card 
sub get_highest_non_pair_card {
    	my ($hand) = @_;
    	my @temp_hand;

    	foreach my $card (@$hand) {
        	if (count_occurrences($hand, $card->{rank}) != 2) {
            		push @temp_hand, $card;
        	}
    	}

    	return unless @temp_hand;

    	my $highest_card = $temp_hand[0];
    	foreach my $card (@temp_hand[1 .. $#temp_hand]) {
        	if ($card->{rank} > $highest_card->{rank}) {
            		$highest_card = $card;
        	}
    	}

    	return $highest_card;
}

# Identifies and returns the index of the kicker card in the hand
sub get_kicker {
    	my ($hand) = @_;
    	my $max_val = 0;
    	my $kicker_index = -1;

    	foreach my $index (0 .. $#$hand) {
        	my $card = $hand->[$index];
        	if (count_occurrences($hand, $card->{rank}) == 1 && $card->{rank} > $max_val) {
            		$max_val = $card->{rank};
            		$kicker_index = $index;
        	}
    	}

    	return $kicker_index;
}

# Gets the rank (value) of the pair in the hand
sub get_pair_rank {
    	my ($hand) = @_;
    	foreach my $card (@$hand) {
        	if (count_occurrences($hand, $card->{rank}) == 2) {
            		return $card->{rank};
        	}
    	}
    	return -1;
}

# Gets the highest rank (value) of pairs in a hand
sub get_highest_pair_rank {
    	my ($hand) = @_;
    	my $highest_pair = -1;

    	foreach my $card (@$hand) {
        	if (count_occurrences($hand, $card->{rank}) == 2 && $card->{rank} > $highest_pair) {
            		$highest_pair = $card->{rank};
        	}
    	}

    	return $highest_pair;
}

# Gets the lowest rank (value) of pairs in a hand
sub get_lowest_pair_rank {
    	my ($hand) = @_;
    	my $lowest_pair = 15;

    	foreach my $card (@$hand) {
        	if (count_occurrences($hand, $card->{rank}) == 2 && $card->{rank} < $lowest_pair) {
            		$lowest_pair = $card->{rank};
        	}
    	}

    	if ($lowest_pair == 15) {
        	return -1;
    	} else {
        	return $lowest_pair;
    	}
}

# Gets the rank of the three-of-a-kind in a hand
sub get_triplet_rank {
    	my ($hand) = @_;

    	foreach my $card (@$hand) {
        	if (count_occurrences($hand, $card->{rank}) == 3) {
            		return $card->{rank};
        	}
    	}

    	return -1;
}

# Gets the rank of the four-of-a-kind in a hand
sub get_quad_rank {
    	my ($hand) = @_;

    	foreach my $card (@$hand) {
        	if (count_occurrences($hand, $card->{rank}) == 4) {
            		return $card->{rank};
        	}
    	}

    	return -1;
}

# Method to compare two hands and determine the winner
sub compare_hands {
    	my ($hand1, $hand2) = @_;
    	my $rank1 = classify_hand($hand1);
    	my $rank2 = classify_hand($hand2);

    	if ($rank1 != $rank2) {
        	if ($rank1 < $rank2) {
            		return -1;
        	} else {
            		return 1;
        	}
    	}

	# Tie-breaking logic based on hand rank
	
	# Case 1: Flush, Straight Flush, Royal Flush
	if ($rank1 == FLUSH || $rank1 == STRAIGHT_FLUSH || $rank1 == ROYAL_FLUSH) {
        	my $highest_card1 = get_highest_card($hand1, 0);
        	my $highest_card2 = get_highest_card($hand2, 0);

        	if ($highest_card1->{rank} != $highest_card2->{rank}) {
            		if ($highest_card1->{rank} < $highest_card2->{rank}) {
                		return -1;
            		} else {
                		return 1;
            	}
        	}
        	if ($highest_card1->{suit} lt $highest_card2->{suit}) {
            		return -1;
        	} else {
            		return 1;
        	}
    	}

	# Case 2: Straight
	if ($rank1 == STRAIGHT) {
        	my $check_ace1 = is_ace_low_straight($hand1);
        	my $check_ace2 = is_ace_low_straight($hand2);

        	my $highest_card1 = get_highest_card($hand1, $check_ace1);
        	my $highest_card2 = get_highest_card($hand2, $check_ace2);
	
        	if ($highest_card1->{rank} != $highest_card2->{rank}) {
            		if ($highest_card1->{rank} < $highest_card2->{rank}) {
                		return -1;
            		} else {
                		return 1;
            		}
        	}
        	if ($highest_card1->{suit} lt $highest_card2->{suit}) {
            		return -1;
        	} else {
            		return 1;
        	}
    	}

	# Case 3: Two Pair
	if ($rank1 == TWO_PAIR) {
        	my $high_pair1 = get_highest_pair_rank($hand1);
        	my $high_pair2 = get_highest_pair_rank($hand2);

        	if ($high_pair1 != $high_pair2) {
            		if ($high_pair1 < $high_pair2) {
                		return -1;
            		} else {
                		return 1;
            		}
        	}

        	my $low_pair1 = get_lowest_pair_rank($hand1);
        	my $low_pair2 = get_lowest_pair_rank($hand2);

        	if ($low_pair1 != $low_pair2) {
            		if ($low_pair1 < $low_pair2) {
                		return -1;
            		} else {
                		return 1;
            		}
        	}

        	my $kicker1 = get_kicker($hand1);
        	my $kicker2 = get_kicker($hand2);
        	if ($hand1->[$kicker1]->{suit} lt $hand2->[$kicker2]->{suit}) {
            		return -1;
        	} else {
            		return 1;
        	}
    	}
	
	# Case 4: Pair
	if ($rank1 == PAIR) {
        	my $pair_rank1 = get_pair_rank($hand1);
        	my $pair_rank2 = get_pair_rank($hand2);

        	if ($pair_rank1 != $pair_rank2) {
            		if ($pair_rank1 < $pair_rank2) {
                		return -1;
            		} else {
                		return 1;
            		}
        	}

        	my $high_val1 = get_highest_non_pair_card($hand1);
        	my $high_val2 = get_highest_non_pair_card($hand2);

        	if ($high_val1->{rank} != $high_val2->{rank}) {
            		if ($high_val1->{rank} < $high_val2->{rank}) {
                		return -1;
            		} else {
                		return 1;
            		}
        	}
        	if ($high_val1->{suit} lt $high_val2->{suit}) {
            		return -1;
        	} else {
            		return 1;
        	}
    	}

	# Case 5: High Card
	if ($rank1 == HIGH_CARD) {
        	my $highest_card1 = get_highest_card($hand1, 0);
        	my $highest_card2 = get_highest_card($hand2, 0);

        	if ($highest_card1->{rank} != $highest_card2->{rank}) {
            		if ($highest_card1->{rank} < $highest_card2->{rank}) {
                		return -1;
            		} else {
                		return 1;
            		}
        	}
        	if ($highest_card1->{suit} lt $highest_card2->{suit}) {
            		return -1;
        	} else {
            		return 1;
        	}
    	}

	# Additional Cases: Three of a Kind, Four of a Kind, Full House
	if ($rank1 == FOUR_OF_A_KIND) {
        	my $quad_rank1 = get_quad_rank($hand1);
        	my $quad_rank2 = get_quad_rank($hand2);
        	if ($quad_rank1 < $quad_rank2) {
            		return -1;
        	} else {
            		return 1;
        	}
    	}

    	if ($rank1 == THREE_OF_A_KIND) {
        	my $triplet_rank1 = get_triplet_rank($hand1);
        	my $triplet_rank2 = get_triplet_rank($hand2);
        	if ($triplet_rank1 < $triplet_rank2) {
            		return -1;
        	} else {
            		return 1;
        	}
    	}

    	if ($rank1 == FULL_HOUSE) {
        	my $triplet_rank1 = get_triplet_rank($hand1);
        	my $triplet_rank2 = get_triplet_rank($hand2);

        	if ($triplet_rank1 != $triplet_rank2) {
            		if ($triplet_rank1 < $triplet_rank2) {
                		return -1;
            		} else {
                		return 1;
            		}
        	}

        	my $pair_rank1 = get_pair_rank($hand1);
        	my $pair_rank2 = get_pair_rank($hand2);
        	if ($pair_rank1 < $pair_rank2) {
            		return -1;
        	} else {
            		return 1;
        	}
    	}

    return 0;
}
